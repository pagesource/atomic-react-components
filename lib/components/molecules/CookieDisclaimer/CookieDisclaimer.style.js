import { css } from 'styled-components';

const disclaimerStyles = css`
  display: flex;
  flex-flow: column;
  border: 1px solid grey;
  padding: ${(props) => props.theme.spacing.UNIT_4};
  animation: ${(props) => props.theme.animations.ANIMATION_SLIDE_UP_LINEAR};
  position: absolute;
  top: 2rem;
  left: 0;
  --webkit-backface-visibility: hidden;
  align-items: flex-end;
  ul {
    list-style: none;
    display: flex;
    li:not(:last-child) {
      margin: 0 ${(props) => props.theme.spacing.UNIT_1} 0 0;
    }
  }
`;

export default disclaimerStyles;
