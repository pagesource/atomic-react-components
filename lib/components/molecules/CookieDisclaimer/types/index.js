// @flow

import type { Node } from 'react';

export type cookieDisclaimerProps = {
  children: Node,
  className?: string,
  acceptBtnText: string,
  rejectBtnText: string,
  cookieDetails?: Object,
  onClick: Function,
};
