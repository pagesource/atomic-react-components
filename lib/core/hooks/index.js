export { default } from './speechRecognition/useSpeechRecognition';
export { useSpeechRecognition } from './speechRecognition/useSpeechRecognition';
