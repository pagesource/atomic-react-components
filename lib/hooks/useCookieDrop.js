/**
 * @param {string} cookieValue -> expected in string format
 * @param {Object} cookieDetails - cookie object to be passed for setting
 * the value which expects name of the cookie and it's expiration time.
 *
 * Sets the cookie according to expiration time.
 */
const setCookie = (cookieValue, cookieDetails) => {
  const { cookieName, expirationDays } = cookieDetails;
  let expiryDate = '';
  if (expirationDays) {
    const date = new Date();
    date.setTime(date.getTime() + expirationDays * 24 * 60 * 60 * 1000);
    expiryDate = `; expires=" ${date.toUTCString()}`;
  }
  document.cookie = `${cookieName}=${cookieValue || ''}${expiryDate}; path=/`;
};

/**
 *
 * @param {object} cookieDetails -> cookie object
 *
 * returns the cookie value corresponding to the name.
 */

const getCookie = (cookieDetails) => {
  const { cookieName } = cookieDetails;
  const regExp = `(^|;) ?${cookieName}=([^;]*)(;|$)`;
  const val = document.cookie.match(regExp);
  return val ? val[2] : null;
};

/**
 *
 * @param {string} name
 *
 * removes the cookie w.r.t the name provided.
 */

const removeCookie = (name) => {
  document.cookie = `${name}=1; expires=1 Jan 1970 00:00:00 GMT;`;
};
export { setCookie, getCookie, removeCookie };
