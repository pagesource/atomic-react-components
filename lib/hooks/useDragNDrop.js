import { useState } from 'react';
import classNames from 'classnames';

const useDragAndDrop = () => {
  const [highlight, setHighlight] = useState(false);

  const fileListToArray = (list) => {
    const array = [];

    for (let i = 0; i < list.length; i += 1) {
      array.push(list.item(i));
    }
    return array;
  };

  const onDragOver = (evt) => {
    evt.preventDefault();
    setHighlight(true);
  };

  const onDragLeave = (e) => {
    e.preventDefault();
    setHighlight(false);
  };

  const onDrop = (event) => {
    event.preventDefault();
    const { files } = event.dataTransfer;
    const array = fileListToArray(files);
    return array;
  };

  const onFilesAdded = (evt) => {
    const { files } = evt.target;
    const array = fileListToArray(files);
    return array;
  };

  const getHighlightClass = (className) => {
    const styleCls = classNames(className, {
      highlight,
    });
    return styleCls;
  };

  return {
    highlight,
    onDragOver,
    onDragLeave,
    onDrop,
    onFilesAdded,
    fileListToArray,
    getHighlightClass,
    setHighlight,
  };
};

export default useDragAndDrop;
