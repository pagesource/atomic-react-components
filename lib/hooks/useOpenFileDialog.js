const useOpenFileDialog = () => {
  const openFileDialog = (fileInput) => {
    fileInput.current.click();
  };
  return {
    openFileDialog,
  };
};

export default useOpenFileDialog;
