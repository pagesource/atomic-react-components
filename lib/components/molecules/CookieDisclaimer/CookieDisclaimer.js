// @flow
import React, { useState } from 'react';
import type { Node, ComponentType } from 'react';
import styled from 'styled-components';

import Button from '../../atoms/Button';
import Icon from '../../atoms/Icon';
// $flow-disable-line eslint-disable-line
import { ReactComponent as IconClose } from '../../../static/images/close.svg';
import type { cookieDisclaimerProps } from './types';
import styles from './CookieDisclaimer.style';
import { setCookie, getCookie } from '../../../hooks/useCookieDrop';

let [isVisible, setIsVisible] = [
  true,
  (val: any): any => {
    return val;
  },
];

/**
 * toggle the visible state of the popup.
 */
export const toggleIsVisibleState = () => {
  setIsVisible(!isVisible);
};

/**
 *
 * @param {Object} cookieDetails
 *
 * setting the value of the cookie when user consent is declined & performing corresponding action.
 */
export const consentRejected = (cookieDetails: any) => {
  setCookie('NO', cookieDetails);
  toggleIsVisibleState();
};

/**
 *
 * @param {Object} cookieDetails
 *
 * setting the value of the cookie when user consent is provided & performing corresponding action.
 */
export const consentAccepted = (cookieDetails: any) => {
  setCookie('YES', cookieDetails);
  toggleIsVisibleState();
};

/**
 *
 * @param {Object} cookieDetails
 *
 * reading the cookie value to conclude the initial state for showing up disclaimer or not.
 */
export const getShowDisclaimer = (cookieDetails: any) => {
  const showDisclaimer = getCookie(cookieDetails);
  return showDisclaimer === null;
};

const CookieDisclaimer = ({
  acceptBtnText,
  rejectBtnText,
  className,
  children,
  cookieDetails,
}: cookieDisclaimerProps): Node => {
  const showDisclaimer = getShowDisclaimer(cookieDetails);
  [isVisible, setIsVisible] = useState(showDisclaimer);
  const disclaimerElement = isVisible ? (
    <div>
      <div className={className} role="alert">
        <Button tertiary onClick={toggleIsVisibleState} aria-label="close Cookie Disclaimer">
          <Icon title="close icon">
            <IconClose />
          </Icon>
        </Button>
        <div>{children}</div>
        <ul>
          <li>
            <Button primary onClick={() => consentAccepted(cookieDetails)}>
              {acceptBtnText}
            </Button>
          </li>
          <li>
            <Button tertiary onClick={() => consentRejected(cookieDetails)}>
              {rejectBtnText}
            </Button>
          </li>
        </ul>
      </div>
    </div>
  ) : null;

  return disclaimerElement;
};

CookieDisclaimer.defaultProps = {
  acceptBtnText: 'Accept',
  className: 'cookie-disclaimer',
  rejectBtnText: 'Reject',
  cookieDetails: {
    cookieKey: 'USER_CONSENT',
    expiryDays: 365,
  },
};

const StyledCookieDisclaimer: ComponentType<cookieDisclaimerProps> = styled(CookieDisclaimer)`
  ${styles};
`;

export default StyledCookieDisclaimer;

export { CookieDisclaimer as CookieDisclaimerVanilla };
