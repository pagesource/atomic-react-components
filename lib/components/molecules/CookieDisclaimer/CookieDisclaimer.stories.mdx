import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import CookieDisclaimer, {CookieDisclaimerVanilla} from './CookieDisclaimer';
import stickyCookieDisclaimer from './CookieDisclaimer.mock';
import Heading from '../../atoms/Heading';
import Para from '../../atoms/Para';
import Anchor from '../../atoms/Anchor';

<Meta title="Molecules/Cookie Disclaimer" component={CookieDisclaimerVanilla} />

### Default
<Preview>
  <Story name="Default">
  <CookieDisclaimer {...stickyCookieDisclaimer}>
      <div>
        <Heading>Privacy and Cookie Policy</Heading>
        <Para>
          We use the cookies to give you the best experience on our website. By continuing,
          you&#39;re agreeing to our use of cookies. we recently updated our policy.
          <Anchor>Learn More</Anchor>
        </Para>
      </div>
    </CookieDisclaimer>
  </Story>
</Preview>

## Props ðŸ”§

<Props of={CookieDisclaimerVanilla} />

