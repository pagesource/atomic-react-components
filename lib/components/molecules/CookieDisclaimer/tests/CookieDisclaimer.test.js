import React from 'react';
import { shallow } from 'enzyme';

import { mountWithTheme } from '../../../../../__mock__/themeMock';
import stickyCookieDisclaimer from '../CookieDisclaimer.mock';
import CookieDisclaimer, {
  CookieDisclaimerVanilla,
  toggleIsVisibleState,
  consentAccepted,
  consentRejected,
  getShowDisclaimer,
} from '../CookieDisclaimer';
import { removeCookie } from '../../../../hooks/useCookieDrop';

const defaultProps = {
  ...stickyCookieDisclaimer,
};
/**
 * getWrapperSetup: to get the vanilla component setup.
 */
const getWrapperSetup = (props = {}) => {
  const finalProps = { ...defaultProps, ...props };
  return shallow(<CookieDisclaimerVanilla {...finalProps} />);
};
/**
 * getMountedWrapperSetup: to get the styled component setup.
 */
const getMountedWrapperSetup = (props = {}) => {
  const finalProps = { ...defaultProps, ...props };
  const vanillaWrapper = <CookieDisclaimer {...finalProps} />;
  return mountWithTheme(vanillaWrapper);
};

describe("Test Suite 1: Check for component and it's child rendered correctly.", () => {
  let wrapper;
  /**
   * beforeEach: method to get wrapper setup every time before each tests run.
   */
  beforeEach(() => {
    const mockSetIsVisible = jest.fn();
    React.useState = jest.fn(() => [true, mockSetIsVisible]);
    wrapper = getWrapperSetup();
  });

  test('should render wrapper correctly for snapshot matching.', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Cookie Disclaimer component rendered correctly.', () => {
    expect(wrapper.find('div')).toHaveLength(3);
  });

  it.skip('should render two buttons for accepting and rejecting the consent', () => {
    expect(wrapper.find('Button')).toHaveLength(3);
  });
});

describe("Test Suite 2: Check for component and it's child for default props correctly.", () => {
  let wrapper;
  /**
   * beforeEach: method to get wrapper setup every time before each tests run.
   */
  beforeEach(() => {
    wrapper = getWrapperSetup();
  });

  it.skip('Button props for button types to be rendered.', () => {
    const firstBtn = wrapper.find('Button');
    expect(firstBtn.at(1).text()).toBe('Accept & Continue');
  });

  it.skip('Button props for button types to be rendered.', () => {
    const firstBtn = wrapper.find('Button');
    expect(firstBtn.at(2).text()).toBe('Decline');
  });
});

describe('Test Suite 3: testing styled component - HOC', () => {
  let wrapper;
  /**
   * beforeEach: method to get wrapper setup every time before each tests run.
   */
  beforeEach(() => {
    wrapper = getMountedWrapperSetup();
  });

  it('should rendered the component properly when styles are being applied.', () => {
    expect(wrapper.html()).not.toBe(null);
  });
});

describe('Test Suite 4: testing functional behaviour of the cookie disclaimer.', () => {
  let wrapper;
  /**
   * beforeEach: method to get wrapper setup every time before each tests run.
   */
  beforeEach(() => {
    wrapper = getMountedWrapperSetup();
  });

  afterEach(() => {
    removeCookie('USER_CONSENT');
  });

  it('setting the isVisible variable to false', () => {
    const mockSetIsVisible = jest.fn();
    const [isVisible, setIsVisible] = [true, jest.fn(() => [false, mockSetIsVisible])];

    toggleIsVisibleState(isVisible, setIsVisible);
    expect(wrapper.html()).not.toBe(null);
  });

  it('accepting the consent and dropping the cookie.', () => {
    const mockSetIsVisible = jest.fn();
    const [isVisible, setIsVisible] = [true, jest.fn(() => [false, mockSetIsVisible])];
    consentAccepted(isVisible, setIsVisible);
    expect(wrapper.html()).not.toBe(null);
  });

  it('rejecting the consent and dropping the cookie.', () => {
    const mockSetIsVisible = jest.fn();
    const [isVisible, setIsVisible] = [true, jest.fn(() => [false, mockSetIsVisible])];
    consentRejected(isVisible, setIsVisible);
    expect(wrapper.html()).not.toBe(null);
  });

  it('getting the disclaimer state for rendering cookie disclaimer.', () => {
    const { cookieDetails } = defaultProps;
    const value = getShowDisclaimer(cookieDetails);
    expect(value).toBe(true);
  });

  it.skip('Accept Button click simulation', () => {
    const btnWrapper = wrapper.find('Button').at(2);
    const acceptBtn = btnWrapper.simulate('click');
    expect(acceptBtn.text()).toBe('Accept & Continue');
  });

  it.skip('Reject Button click simulation', () => {
    const btnWrapper = wrapper.find('Button').at(4);
    const rejectBtn = btnWrapper.simulate('click');
    expect(rejectBtn.text()).toBe('Decline');
  });
});
