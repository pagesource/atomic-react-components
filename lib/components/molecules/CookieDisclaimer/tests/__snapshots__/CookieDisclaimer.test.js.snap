// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Suite 1: Check for component and it's child rendered correctly. should render wrapper correctly for snapshot matching. 1`] = `
<div>
  <div
    className="cookie-disclaimer"
    role="alert"
  >
    <Button__StyledButton
      aria-label="close Cookie Disclaimer"
      onClick={[Function]}
      tertiary={true}
    >
      <Icon__StyledIcon
        title="close icon"
      >
        <ReactComponent />
      </Icon__StyledIcon>
    </Button__StyledButton>
    <div />
    <ul>
      <li>
        <Button__StyledButton
          onClick={[Function]}
          primary={true}
        >
          Accept & Continue
        </Button__StyledButton>
      </li>
      <li>
        <Button__StyledButton
          onClick={[Function]}
          tertiary={true}
        >
          Decline
        </Button__StyledButton>
      </li>
    </ul>
  </div>
</div>
`;
